### ISSUE TITLE
mse_loss reduction='none' is ignored when required_grads is True

### ISSUE THREAD

**Issue by akhti** [pytorch/pytorch#10009]
## Issue description

If a tensor with requires_grad=True is passed to mse_loss, then the loss is reduced even if reduction is none.
Appeared in Pytorch 0.4.1.

## Code example

```(python)
import torch
x = torch.zeros((4, 5, 2))
print('Good', torch.nn.functional.mse_loss(x, x, reduction='none').shape)
x.requires_grad = True
print('Bad', torch.nn.functional.mse_loss(x, x, reduction='none').shape)
```
Outputs:
```
Good torch.Size([4, 5, 2])
Bad torch.Size([])
```
## System Info
```
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
```

**Comment by zou3519** [comment]
cc @li-roy

**Comment by li-roy** [comment]
PR is out for this.

@akhti  FYI, this bug only happens when the target tensor requires grad. If only input requires grad, the bug won't happen.

**Comment by mjschoeberl** [comment]
Unfortunately the reduction of the mse_loss is still performed despite setting reduction='none' if both, input and target, requires_grad is True.

    input = torch.randn(10,2, requires_grad =True)
    target = torch.torch.ones(10,2,requires_grad=True)
    loss = torch.nn.functional.mse_loss(input, target, reduction='none')
    print loss.shape

In case only the input requires a gradient and not the target, the output shape of the loss is correct.

Has anyone observed the same phenomenon? Or am I abusing the function by having target and input requiring a gradient? 
Apologizes in case my comment is obsolete. Thank you for your help.

Python: 2.7
PyTorch: 0.4.1 (CPU) Updated on 10/14/2018

**Comment by li-roy** [comment]
I think you're on an older version, @mjschoeberl. I can't reproduce that on master.

**Comment by mjschoeberl** [comment]
Many thanks @li-roy, with 1.0.0.dev20181018 it works smoothly.
